core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -460,25 +493,213 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
         dagViz ++
         maybeExpandDagViz ++
         showAdditionalMetrics ++
+        makeTimeline(stageData.taskData.values.toSeq, currentTime) ++
         <h4>Summary Metrics for {numCompleted} Completed Tasks</h4> ++
         <div>{summaryTable.getOrElse("No tasks have reported metrics yet.")}</div> ++
         <h4>Aggregated Metrics by Executor</h4> ++ executorTable.toNodeSeq ++
         maybeAccumulableTable ++
         <h4>Tasks</h4> ++ taskTable
-
       UIUtils.headerSparkPage(stageHeader, content, parent, showVisualization = true)
     }
   }
 
+  def makeTimeline(tasks: Seq[TaskUIData], currentTime: Long): Seq[Node] = {
+    val executorsSet = new HashSet[(String, String)]
+    var minLaunchTime = Long.MaxValue
+    var maxFinishTime = Long.MinValue
+    var numEffectiveTasks = 0
+
+    val executorsArrayStr = tasks.filter { taskUIData =>
+      val taskInfo = taskUIData.taskInfo
+      taskInfo.successful || taskInfo.running || taskInfo.failed
```

Comment:
The "UNKNOWN" status cannot ever occur.  Using boolean algebra for what you've written:

successful || running || failed
expands to: (finished & !failed) || !finished || failed
is equivalent to: (finished & !failed) || !finished || (failed & !finished) || (failed & finished)
the first and last terms can be combined to get: finished || !finished || (failed & !finished)
which is trivially true (because either finished or !finished will be true!)

(separately, I'll submit a PR later to remove the "UNKNOWN" state)